[OSDOCS-2604] Document ShiftStack cloud provider option configuration (#36088)
openshift/openshift-docs: modules/installation-osp-setting-cloud-provider-options.adoc

For a complete list of cloud provider configuration parameters, see link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].

.Prerequisites

* Have manifest files generated for your cluster.
** If you don't have manifest files, generate them by running `openshift-install --dir <destination_directory> create manifests` from a command line.

.Procedure

. In a text editor, open the cloud-provider configuration manifest file. For example:
+
[source,terminal]
----
$ vi openshift/manifests/cloud-provider-config.yaml
----

. Modify the options based on link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].
+
Configuring Octavia for load balancing is a common case. For example:
+
[source,yaml]
----
#...
[LoadBalancer]
use-octavia=true <1>
lb-provider = "amphora" <2>
floating-network-id="d3deb660-4190-40a3-91f1-37326fe6ec4a"<3>
floating-subnet-id="4a2fe995-84d4-4eac-8bcf-c5624f63dba0"
subnet-id="e3d26bd3-0b6b-4c4d-b184-71c5be59b8b2"
monitor-delay=60s
monitor-timeout=30s

> monitor-max-retries=5

Reviewer: You should also remove these options that no longer make sense without `create-monitor=true`

Author: ```suggestion
```

For a complete list of cloud provider configuration parameters, see link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].

.Prerequisites

* Have manifest files generated for your cluster.
** If you don't have manifest files, generate them by running `openshift-install --dir <destination_directory> create manifests` from a command line.

.Procedure

. In a text editor, open the cloud-provider configuration manifest file. For example:
+
[source,terminal]
----
$ vi openshift/manifests/cloud-provider-config.yaml
----

. Modify the options based on link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].
+
Configuring Octavia for load balancing is a common case. For example:
+
[source,yaml]
----
#...
[LoadBalancer]
use-octavia=true <1>
lb-provider = "amphora" <2>
floating-network-id="d3deb660-4190-40a3-91f1-37326fe6ec4a"<3>
floating-subnet-id="4a2fe995-84d4-4eac-8bcf-c5624f63dba0"
subnet-id="e3d26bd3-0b6b-4c4d-b184-71c5be59b8b2"

> create-monitor=true

Reviewer: The `create-monitor=true` option was not tested by QE as far as I can tell, I'd suggest removing it from the example until we confirmed it works as expected.

Author: ```suggestion
```

monitor-delay=60s
monitor-timeout=30s
monitor-max-retries=5
#...
----
<1> This property enables Octavia integration.
<2> This property sets the Octavia provider that your load balancer uses. It accepts `"ovn"` or `"amphora`" as values. If you choose to use OVN, you must also set `lb-method` to `SOURCE_IP_PORT`.
<3> This property is required if you want to use multiple external networks with your cluster. The cloud provider creates floating IP addresses on the network that is specified here.

> 
> . Save the changes to the file and proceed with installation.
> 

Author: FYI @dulek from https://github.com/openshift/openshift-docs/pull/36088/files/80343fcc4cceca59c81751af820d6ca0979062d8#r707485144

```suggestion

[IMPORTANT]
====
Prior to saving your changes, verify that the file is structured correctly. Clusters may fail if properties are not placed in the appropriate section.
====

. Save the changes to the file and proceed with installation.

```

Reviewer 2: Looks great!

[TIP]
====
You can update your cloud provider configuration after you run the installer. On a command line, run:

[source,terminal]
----
$ oc edit configmap -n openshift-config cloud-provider-config
----

> After you save your changes, your cluster will take some time to reconfigure itself. The process is complete if none of your nodes have a `SchedulingDisabled` status.

Reviewer 2: I'd add a warning that it is important to double-check the file structure before saving it. We've seen clusters breaking if e.g. options were put in wrong sections.

//Module included in the following assemblies:
//
// * installing/installing_openstack/installing-openstack-installer.adoc
// * installing/installing_openstack/installing-openstack-installer-custom.adoc
// * installing/installing_openstack/installing-openstack-installer-kuryr.adoc
// * installing/installing_openstack/installing-openstack-installer-user.adoc

[id="installation-osp-setting-cloud-provider-options_{context}"]
= Setting cloud provider options

Optionally, you can edit the cloud provider configuration for your cluster. The cloud provider configuration controls how {product-title} interacts with {rh-openstack-first}.

For a complete list of cloud provider configuration parameters, see link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].

.Prerequisites

> * Have manifest files generated for your cluster.

Author: Done in 07fa501914a5247a531e42706b2c430fe3a41262

> ** If you don't have manifest files, generate them by running `openshift-install --dir <destination_directory> create manifests` from a command line.

Author: ```suggestion
** If you do not have manifest files, generate them by running `openshift-install --dir <destination_directory> create manifests` from a command line.
```

.Procedure

. In a text editor, open the cloud-provider configuration manifest file. For example:
+
[source,terminal]
----
$ vi openshift/manifests/cloud-provider-config.yaml
----

. Modify the options based on link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].
+
Configuring Octavia for load balancing is a common case. For example:
+
[source,yaml]
----
#...
[LoadBalancer]
use-octavia=true <1>
lb-provider = "amphora" <2>
floating-network-id="d3deb660-4190-40a3-91f1-37326fe6ec4a"<3>
floating-subnet-id="4a2fe995-84d4-4eac-8bcf-c5624f63dba0"

> subnet-id="e3d26bd3-0b6b-4c4d-b184-71c5be59b8b2"

Reviewer 2: Hm, we should probably either explain this or just omit this option (cloud-provider will autodetect it). Basically it's about in which subnet to create the load balancers. We've never tested anything other than the subnet used by the nodes, so I'd rather remove this line.

@iamemilio: I just realized that we haven't checked how the autodetection works if you have nodes in another subnet. There may be dragons.

Author: @eurijon No qualms with removing this line? 

```suggestion
```

Reviewer 3: It can be removed as it's autodetected by the cloud-provider

Author: TY.

#...
----
<1> This property enables Octavia integration.
<2> This property sets the Octavia provider that your load balancer uses. It accepts `"ovn"` or `"amphora`" as values. If you choose to use OVN, you must also set `lb-method` to `SOURCE_IP_PORT`.
<3> This property is required if you want to use multiple external networks with your cluster. The cloud provider creates floating IP addresses on the network that is specified here.
+
[IMPORTANT]
====

> Prior to saving your changes, verify that the file is structured correctly. Clusters may fail if properties are not placed in the appropriate section.

Author: ```suggestion
Prior to saving your changes, verify that the file is structured correctly. Clusters might fail if properties are not placed in the appropriate section.
```

====

> . Save the changes to the file and proceed with installation.

Reviewer 3: In the preview this sentence is preceded by the index n.1 when it should n.3 I think

> <1> This property enables Octavia integration.
> <2> This property sets the Octavia provider that your load balancer uses. It accepts `"ovn"` or `"amphora`" as values. If you choose to use OVN, you must also set `lb-method` to `SOURCE_IP_PORT`.
> <3> This property is required if you want to use multiple external networks with your cluster. The cloud provider creates floating IP addresses on the network that is specified here.
> 
> [IMPORTANT]
> ====
> Prior to saving your changes, verify that the file is structured correctly. Clusters may fail if properties are not placed in the appropriate section.
> ====
> 
> . Save the changes to the file and proceed with installation.

Author: ```suggestion
<1> This property enables Octavia integration.
<2> This property sets the Octavia provider that your load balancer uses. It accepts `"ovn"` or `"amphora`" as values. If you choose to use OVN, you must also set `lb-method` to `SOURCE_IP_PORT`.
<3> This property is required if you want to use multiple external networks with your cluster. The cloud provider creates floating IP addresses on the network that is specified here.
+
[IMPORTANT]
====
Prior to saving your changes, verify that the file is structured correctly. Clusters may fail if properties are not placed in the appropriate section.
====

. Save the changes to the file and proceed with installation.
```

> [TIP]

Reviewer 4: This should get a preceding `+` for indentation.

Author: Done in 57cb23e87ad98ea28dab195fd03a43869c8f9d9f

Optionally, you can edit the cloud provider configuration for your cluster. The cloud provider configuration controls how {product-title} interacts with {rh-openstack-first}.

For a complete list of cloud provider configuration parameters, see link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].

.Procedure

. If you have not already generated manifest files for your cluster, generate them by running the following command:
+
[source,terminal]
----
$ openshift-install --dir <destination_directory> create manifests
----

. In a text editor, open the cloud-provider configuration manifest file. For example:
+
[source,terminal]
----
$ vi openshift/manifests/cloud-provider-config.yaml
----

. Modify the options based on link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].
+
Configuring Octavia for load balancing is a common case. For example:
+
[source,yaml]
----
#...
[LoadBalancer]
use-octavia=true <1>
lb-provider = "amphora" <2>
floating-network-id="d3deb660-4190-40a3-91f1-37326fe6ec4a"<3>

> floating-subnet-id="4a2fe995-84d4-4eac-8bcf-c5624f63dba0"

Reviewer: `floating-subnet-id` is not a valid option according to https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf.

Setting it causes errors in openshift-kube-controller-manager:
```
F0929 11:19:55.041313       1 controllermanager.go:246] error building controller context: cloud provider could not be initialized: could not init cloud provider "openstack": warning:                                                      
can't store data at section "LoadBalancer", variable "floating-subnet-id"
```

Author: Gone!
```suggestion
```
